### ISSUE TITLE
Inconsistent autocomplete behavior

### ISSUE THREAD

**Issue by basaran** [sublimelsp/LSP#745]
### Description
When you invoke sublime's autocomplete without any string, a consequent invocation will not show any LSP information. Please see the gif.

![lsp-php](https://user-images.githubusercontent.com/30809170/66140454-251b1180-e5d0-11e9-8481-342dee2e3be6.gif)

### Steps to reproduce

1. Type Req inside a method, and invoke autocomplete via shortcut.
2. Observe LSP is providing the use statement for Request class.
3. Delete everything, go to method, just invoke autocomplete via shortcut and type Req. Nothing is provided by LSP.
4. Move the block down by a line, type Req and invoke autocomplete. Notice LSP is working again.

Unless the line is moved, or the parenthesizes removed, LSP will not work again, defaulting to sublime autocomplete only.

### Expected behavior

LSP should always show the hints when the autocomplete is invoked.

### Environment

* Build: 3210
* Operating system and version: Linux Arch
* [Linux] Desktop Environment and/or Window Manager: Awesomewm

Also submitted to:
https://github.com/SublimeTextIssues/Core/issues/3005

**Comment by tomv564** [comment]
Hi, thanks for the detailed report. I feel a bit unclear from the extra steps before/after step 3. Could it be that triggering the completion shortcut _before_ typing `Req` that causes the issue?

Could you try two things:
* Set `log_debug` and `log_payloads` to True, restart Sublime and share the sublime console log for only the "trigger completion and type Req" part of the repro?
* See if the behaviour is the same with the `only_show_lsp_completions` LSP package setting set to True?

**Comment by basaran** [comment]
Hello, I made you a new screen capture. My autocomplete is set to ctrl+space.

![lsp](https://user-images.githubusercontent.com/30809170/66172105-9da5c080-e618-11e9-8064-4b9dffe4d3fa.gif)

Making another one with the only_show_lsp_completions shortly, and with the log as well.

**Comment by basaran** [comment]
Here's another capture with  lsp only set to true:

![lsp-only](https://user-images.githubusercontent.com/30809170/66172924-c8454880-e61b-11e9-96f6-1fcc15b23e08.gif)

and this is the log:

[lsp_only.txt](https://github.com/tomv564/LSP/files/3688513/lsp_only.txt)

**Comment by tomv564** [comment]
That's much clearer, thanks.
In the logs I see 5 requests for completions to intelephense, with responses [data, empty, data, empty, data].

In your screenshot you can also this response in the console when you are pressing <kbd>ctrl</kbd><kbd>space</kbd> after deleting the first "Req": 
```
LSP:  --> textDocument/completion
LSP:      {'items': [], 'isIncomplete': False}
```
Once intelephense sends an empty response at a given location, we will cache it until you edit the document.

If you can get some logging from the server then we could know for sure, but I suspect intelephense thinks the document is in a different state than your editor shows.

**Comment by basaran** [comment]
I noticed that too, it seems invoking the autocomplete with no input does not trigger the LSP hints. After that, even if you type something in, it seems think nothing has changed, unless you remove the brackets or shift the line to another line and then type in something. 

So, you think the problem is with lsp-intelephense?

**Comment by tomv564** [comment]
Yes, most servers return completion items if the completion request has no characters preceding the cursor. Intelephense's choice not to is the cause for what you're seeing with LSP.

Have you tried VS Code with intelephense? If the behaviour there is better then we can have a closer look at their handling of empty responses.

**Comment by basaran** [comment]
Tested with VS Code + Intelephense. It has the same behavior, if you invoke the code complete with nothing typed in, it only gives you local suggestions. However, it's able to recover without moving things around. I did it twice so it can be seen easier.

Capture:

![vscode](https://user-images.githubusercontent.com/30809170/66333245-80b50a00-e904-11e9-8267-53a069f7adfe.gif)

**Comment by tomv564** [comment]
Re-opening because I reverted the fix for causing extra completion requests on all servers.

I think it can be fixed by differentiating between `self.completions` is `[]` (server replied no results, try again with more chars) and `self.completions` is `None` (waiting for reply, don't send more requests)

**Comment by rwols** [comment]
I propose to wontfix for ST build 3211. For later builds of ST, we'll have the completion logic almost entirely rewritten. Including having way less state, so what you're asking for is probably fixed alongside the changes.
